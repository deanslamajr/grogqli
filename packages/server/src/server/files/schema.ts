import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
