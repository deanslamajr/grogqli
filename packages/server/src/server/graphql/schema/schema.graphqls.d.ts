/* 19a699f8a8761279db5b06650f17155319410b52
 * This file is automatically generated by graphql-let. */

import { GraphQLResolveInfo } from 'graphql';
export declare type Maybe<T> = T | null;
export declare type Exact<T extends {
    [key: string]: unknown;
}> = {
    [K in keyof T]: T[K];
};
/** All built-in and custom scalars, mapped to their actual values */
export declare type Scalars = {
    ID: string;
    String: string;
    Boolean: boolean;
    Int: number;
    Float: number;
};
export declare type Recording = {
    __typename?: 'Recording';
    replaceThisBullshxt: Scalars['String'];
};
export declare type Query = {
    __typename?: 'Query';
    getRecordings: Array<Maybe<Recording>>;
};
export declare type ResolverTypeWrapper<T> = Promise<T> | T;
export declare type LegacyStitchingResolver<TResult, TParent, TContext, TArgs> = {
    fragment: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type NewStitchingResolver<TResult, TParent, TContext, TArgs> = {
    selectionSet: string;
    resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export declare type StitchingResolver<TResult, TParent, TContext, TArgs> = LegacyStitchingResolver<TResult, TParent, TContext, TArgs> | NewStitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | StitchingResolver<TResult, TParent, TContext, TArgs>;
export declare type ResolverFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => Promise<TResult> | TResult;
export declare type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => AsyncIterator<TResult> | Promise<AsyncIterator<TResult>>;
export declare type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<{
        [key in TKey]: TResult;
    }, TParent, TContext, TArgs>;
    resolve?: SubscriptionResolveFn<TResult, {
        [key in TKey]: TResult;
    }, TContext, TArgs>;
}
export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
    subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
    resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}
export declare type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> = SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs> | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;
export declare type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> = ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>) | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;
export declare type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (parent: TParent, context: TContext, info: GraphQLResolveInfo) => Maybe<TTypes> | Promise<Maybe<TTypes>>;
export declare type IsTypeOfResolverFn<T = {}> = (obj: T, info: GraphQLResolveInfo) => boolean | Promise<boolean>;
export declare type NextResolverFn<T> = () => Promise<T>;
export declare type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (next: NextResolverFn<TResult>, parent: TParent, args: TArgs, context: TContext, info: GraphQLResolveInfo) => TResult | Promise<TResult>;
/** Mapping between all available schema types and the resolvers types */
export declare type ResolversTypes = {
    Recording: ResolverTypeWrapper<Recording>;
    String: ResolverTypeWrapper<Scalars['String']>;
    Query: ResolverTypeWrapper<{}>;
    Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
};
/** Mapping between all available schema types and the resolvers parents */
export declare type ResolversParentTypes = {
    Recording: Recording;
    String: Scalars['String'];
    Query: {};
    Boolean: Scalars['Boolean'];
};
export declare type RecordingResolvers<ContextType = any, ParentType extends ResolversParentTypes['Recording'] = ResolversParentTypes['Recording']> = {
    replaceThisBullshxt?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
    __isTypeOf?: IsTypeOfResolverFn<ParentType>;
};
export declare type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
    getRecordings?: Resolver<Array<Maybe<ResolversTypes['Recording']>>, ParentType, ContextType>;
};
export declare type Resolvers<ContextType = any> = {
    Recording?: RecordingResolvers<ContextType>;
    Query?: QueryResolvers<ContextType>;
};
/**
 * @deprecated
 * Use "Resolvers" root object instead. If you wish to get "IResolvers", add "typesPrefix: I" to your config.
 */
export declare type IResolvers<ContextType = any> = Resolvers<ContextType>;

          
// This is an extra code in addition to what graphql-codegen makes.
// Users are likely to use 'graphql-tag/loader' with 'graphql-tag/schema/loader'
// in webpack. This code enables the result to be typed.
import { DocumentNode } from 'graphql'
export default typeof DocumentNode
