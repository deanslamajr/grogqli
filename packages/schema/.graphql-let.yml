schema: 'src/schema/**/*.graphql'
documents: 'src/queries/**/*.graphql'
plugins:
  - typescript
  - typescript-operations
  - typed-document-node
  - typescript-resolvers
schemaEntrypoint: 'src/schema/schema.graphql'
respectGitIgnore: true
cacheDir: src/generated
config:
  avoidOptionals: true
  useIndexSignature: true
