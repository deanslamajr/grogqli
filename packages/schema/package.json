{
  "version": "0.4.0",
  "license": "MIT",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=10"
  },
  "watch": {
    "build:types": {
      "patterns": [
        "src/schema",
        "src/queries"
      ],
      "ignore": [
        "src/generated",
        "src/queries/*.d.ts",
        "dist"
      ],
      "extensions": [
        "graphql"
      ],
      "quiet": false
    }
  },
  "scripts": {
    "build:types": "graphql-let",
    "watch:build:types": "npm-watch build:types",
    "start:dev": "concurrently --raw npm:watch:build:types npm:watch:tsdx",
    "watch:tsdx": "tsdx watch --verbose",
    "build:publish": "tsdx build",
    "test": "tsdx test --passWithNoTests",
    "lint": "tsdx lint --fix",
    "prepare": "tsdx build",
    "size": "size-limit",
    "analyze": "size-limit --why"
  },
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "husky": {
    "hooks": {
      "pre-commit": "tsdx lint"
    }
  },
  "prettier": {
    "printWidth": 80,
    "semi": true,
    "singleQuote": true,
    "trailingComma": "es5"
  },
  "publishConfig": {
    "access": "public"
  },
  "name": "@grogqli/schema",
  "author": "Dean Slama",
  "module": "dist/schema.esm.js",
  "size-limit": [
    {
      "path": "dist/schema.cjs.production.min.js",
      "limit": "10 KB"
    },
    {
      "path": "dist/schema.esm.js",
      "limit": "10 KB"
    }
  ],
  "devDependencies": {
    "@graphql-codegen/cli": "^1.17.8",
    "@graphql-codegen/plugin-helpers": "^1.17.8",
    "@graphql-codegen/typed-document-node": "^1.17.9",
    "@graphql-codegen/typescript": "^1.17.9",
    "@graphql-codegen/typescript-operations": "^1.17.8",
    "@graphql-codegen/typescript-resolvers": "^1.17.10",
    "@graphql-typed-document-node/core": "^3.1.0",
    "@size-limit/preset-small-lib": "^4.6.0",
    "concurrently": "^5.3.0",
    "graphql-let": "^0.16.1",
    "husky": "^4.3.0",
    "npm-watch": "^0.7.0",
    "prettier": "^2.1.2",
    "rollup-plugin-graphql-tag": "^0.1.0",
    "size-limit": "^4.6.0",
    "tsdx": "^0.14.0",
    "tslib": "^2.0.1",
    "typescript": "^4.0.3",
    "wait-on": "^5.2.0"
  },
  "dependencies": {
    "graphql": "^15.3.0"
  }
}
