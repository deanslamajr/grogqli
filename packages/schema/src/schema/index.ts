import { typeDefs as sharedTypes } from './shared';
import { typeDefs as queryTypes } from './queries';
import { typeDefs as mutationTypes } from './mutations';

export const typeDefs = [
  ...queryTypes,
  ...mutationTypes,
  ...sharedTypes,
];