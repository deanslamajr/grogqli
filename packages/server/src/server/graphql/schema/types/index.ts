import { join } from 'path'
import { loadFilesSync } from '@graphql-tools/load-files'
import { mergeTypeDefs } from '@graphql-tools/merge'

const loadedFiles = loadFilesSync(join(process.cwd(), '**/*.graphqls'))
export const typeDefs = mergeTypeDefs(loadedFiles)