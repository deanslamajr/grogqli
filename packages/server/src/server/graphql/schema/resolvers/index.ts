import {Resolvers} from '@grogqli/schema';

import { queryResolver } from './queries';
import {mutationResolver } from './mutations';
import {subscriptionResolver} from './subscriptions';

export const resolvers: Partial<Resolvers> = {
  Query: queryResolver,
  Mutation: mutationResolver,
  Subscription: subscriptionResolver
};
