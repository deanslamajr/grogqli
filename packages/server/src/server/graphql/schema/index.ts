import { makeExecutableSchema } from '@graphql-tools/schema';
import { typeDefs } from '@grogqli/schema';

import { resolvers } from './resolvers';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
