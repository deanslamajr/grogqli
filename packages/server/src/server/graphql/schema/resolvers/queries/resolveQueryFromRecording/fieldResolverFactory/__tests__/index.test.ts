import { fieldResolverFactory } from '..';
import schema from '../../../../__tests__/grogqli/schemas/someSchemaId/schema.json';
import operationsData from '../../../../__tests__/grogqli/schemas/someSchemaId/operations.json';

describe('fieldResolverFactory', () => {
  describe('returns a function', () => {
    it('should handle the root field case', async () => {
      const resolveField = fieldResolverFactory({
        schema: schema.introspectionQuery,
        operationsData,
        parentTypeName: schema.introspectionQuery.__schema.queryType.name,
        fieldName: 'needToFixTheSchema',
      });

      const info = {
        operation: {
          name: {
            value: 'someOpName',
          },
        },
      };

      const actual = await resolveField(
        {},
        {},
        {
          schemaId: 'someSchemaId',
          runTimeVariables: { grogqli: { workflowId: 'someWorkFlowId' } },
        },
        info
      );

      expect(actual).toMatchSnapshot();
    });

    xit('should return the type instance', () => {
      const actual = resolveField(undefined, {}, {}, info);
      expect(actual).toMatchSnapshot();
    });

    describe('if type instance doesnt exist', () => {
      xit('should', () => {});
    });
  });
});
