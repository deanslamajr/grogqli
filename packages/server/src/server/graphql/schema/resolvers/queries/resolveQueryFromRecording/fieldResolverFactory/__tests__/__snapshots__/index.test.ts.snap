// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation an unnamed operation should throw 1`] = `"TODO handle unnamed query case"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if a type file cannot be found for the given root type should throw 1`] = `"TODO Handle case where type file does not exist. typeId:NonExistentRootType"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if a workflow file cannot be found for the given workflowId should throw 1`] = `"TODO handle case where a workflow file is not found for the given workflowId. workflowId:nonexistentWorkflowId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the operations file does not have an recording for the given operation should throw 1`] = `"TODO handle case where a recording does not exist for the given opName. opName:unrecordedOpName"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the root type file does not have a recording instance for the given recordingId should throw 1`] = `"TODO Handle case where type file does not have a recording for the given id. typeId:someTypeId recordingId:nonExistentRecordingId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the root type recording does not have an entry for the given fieldName should throw 1`] = `"TODO handle case where recording doesn't have a value for the given field. fieldName:nonExistentFieldName"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the workflow file does not have an operation recording for the given operation should throw 1`] = `"TODO handle case where the workflow file does not have a recording for the given opId. opId:nonExistentOpId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case should return the value associated with the given fieldName 1`] = `"either a value or a typeInstanceId"`;
